otlp/queue_manager.go

```diff
@@ -394,78 +406,54 @@ func (t *QueueManager) reshardLoop() {
 	}
 }
 
+// reshard changes the number of shards in `t.shards`.  It does not
+// adjust `t.numShards`.  This blocks until the shard number change
+// is complete, but does not wait for stopped shards to exit.
 func (t *QueueManager) reshard(n int) {
-	// Note: There is no requirement that points be written in
-	// order stated for OTLP.  We will start a new set of shards
-	// before stopping the old one.
-
-	newShards := t.newShardCollection(n)
-	newShards.start()
-
+	if n < 1 {
+		// Do not remove all running shards!
+		return
+	}
 	t.shardsMtx.Lock()
 	defer t.shardsMtx.Unlock()
 
-	oldShards := t.shards
-	t.shards = newShards
-
-	oldShards.stop()
-}
-
-type shard struct {
-	queue chan retrieval.SizedMetric
-}
-
-func newShard(cfg promconfig.QueueConfig) shard {
-	return shard{
-		queue: make(chan retrieval.SizedMetric, cfg.Capacity),
-	}
-}
-
-type shardCollection struct {
-	qm     *QueueManager
-	shards []shard
-	done   chan struct{}
-}
-
-func (t *QueueManager) newShardCollection(numShards int) *shardCollection {
-	shards := make([]shard, numShards)
-	for i := 0; i < numShards; i++ {
-		shards[i] = newShard(t.cfg)
+	for len(t.shards) < n {
+		shard := t.newShard()
+		go t.runShard(shard)
+		t.shards[shard] = struct{}{}
 	}
-	s := &shardCollection{
-		qm:     t,
-		shards: shards,
-		done:   make(chan struct{}),
+	for len(t.shards) > n {
```

Comment:
are there tests that already exercise the usecase of downsizing the number of shards?

Reply:
There were no pre-existing tests for resharding-down. I've added a very small test (around lines 300 in `_test.go`). A number of Stackdriver-proto tests were dropped very early in this repo's history, which may have covered that.

Reply:
Since you asked: it was a bit alarming to make all of this change and have nothing break. I had forgotten to `delete()` from the map. Nothing broke, until I wrote the tiny new test.